### PR TITLE
Extract volume interaction to a volumes service

### PR DESCRIPTION
**Pull request by cpuguy83**
This cleans up some of the package API's used for interacting with volumes, and simplifies management.

This is built on top of #36637.

I've done a lot of manual testing on this, but would like to add some more tests especially around the filtering logic, which is also now consolidated to the volume service.

### CODE REVIEW

File: volume/service/service.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+	"github.com/docker/docker/pkg/idtools"
+	"github.com/docker/docker/pkg/plugingetter"
+	"github.com/docker/docker/volume"
+	"github.com/docker/docker/volume/drivers"
+	"github.com/docker/docker/volume/local"
+	"github.com/pkg/errors"
+)
+
+type vs interface {
```
Comment (dnephin) [inline volume/service/service.go]: This seems to be a very different abstraction from what we are calling services currently (ex: ImageService, and RegistryService) which implement higher level logic.

Most of this interface is implemented by the `VolumeStore`, right? 

I would think a `VolumeService` would be a type that implements `api/server/router/volume.Backend`, as well as whatever is used by other services. That way we remove the need for a monolithic daemon package.

I like the changes in this PR, but it seems that we're losing the ability to create a VolumeService to replace the daemon methods for volumes.

Could this continue to be a `VolumeStore`, and we split out a service later?

Comment (cpuguy83) [reply volume/service/service.go]: We could call it that still... but that (replace daemon methods) is indeed the intent, it's just yet more change to add in this large PR, of course can add a commit to make that happen.

Comment (cpuguy83) [reply volume/service/service.go]: Made this change.

**Comment by cpuguy83** [comment]
Wow this is all green.

**Review by vdemeester — APPROVED** [review]
LGTM :frog:

File: daemon/create_unix.go [hunk 2]
```diff
@@ -11,8 +11,10 @@ import (
 	mounttypes "github.com/docker/docker/api/types/mount"
 	"github.com/docker/docker/container"
 	"github.com/docker/docker/pkg/stringid"
+	volumeopts "github.com/docker/docker/volume/service/opts"
 	"github.com/opencontainers/selinux/go-selinux/label"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
```
Comment (thaJeztah) [inline daemon/create_unix.go]: should this be "context" now?

File: daemon/mounts.go [hunk 3]
```diff
@@ -6,7 +6,8 @@ import (
 
 	mounttypes "github.com/docker/docker/api/types/mount"
 	"github.com/docker/docker/container"
-	volumestore "github.com/docker/docker/volume/store"
+	volumesservice "github.com/docker/docker/volume/service"
+	"golang.org/x/net/context"
```
Comment (thaJeztah) [inline daemon/mounts.go]: "context"

File: volume/service/service.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+// GetDriverList gets the list of registered volume drivers
+func (s *VolumesService) GetDriverList() []string {
+	return s.ds.GetDriverList()
+}
+
+// Create creates a volume
+func (s *VolumesService) Create(ctx context.Context, name, driverName string, opts ...opts.CreateOption) (*types.Volume, error) {
+	if name == "" {
+		name = stringid.GenerateNonCryptoID()
```
Comment (thaJeztah) [inline volume/service/service.go]: Wondering if we should start considering giving volumes an ID, and move name to metadata (with`volume rename` in mind). Existing volumes could use name as ID (as they're already unique)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
nice! gave this a first look, saw some nits

**Comment by cpuguy83** [comment]
Updated and all green.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
ping @anusha-ragunathan PTAL

**Comment by zq-david-wang** [comment]
@cpuguy83  unit tests TestServiceCreate/TestServiceGet may fail
https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/21046/console
> 13:35:39 --- FAIL: TestServiceCreate (0.05s)
> 13:35:39 	service_test.go:38: assertion failed: 
> 13:35:39 		{*types.Volume}.CreatedAt:
> 13:35:39 			-: "2018-06-05T13:35:12Z"
> 13:35:39 			+: "2018-06-05T13:35:13Z"
> 13:35:39 		
> 13:35:39 --- FAIL: TestServiceGet (0.08s)
> 13:35:39 	service_test.go:144: assertion failed: 
> 13:35:39 		{*types.Volume}.CreatedAt:
> 13:35:39 			-: "2018-06-05T13:35:12Z"
> 13:35:39 			+: "2018-06-05T13:35:13Z"
> 13:35:39

**Pull request merged**